### ISSUE TITLE
[BUG] no loss masking

### ISSUE THREAD

**Issue by EdanToledo** [instadeepai/Mava#870]
### Multi-agent loss masking
The bug is present in PPO and in general is a side effect of the way reverb stores data. I'm just wondering if anyone has a potential solution as I believe it is quite a structural problem. So agents should have their loss masked according to whether they were actually acting or not. Having the discounts sequence is almost a solution except agents should train on the last step i.e the first time when discount becomes zero. The problem is that due to the multi-agent nature of mava and the way data is stored, it's possible that entire trajectories/sequences of one specific agent is actually null (This happens in PCB environment) and should be masked out e.g if they died in the beginning of the episode. I would imagine this is highly important as certain loss functions can give a loss even when zero padding is used and also training on zeros might be highly undesirable. Beyond padded data, we don't know how environments handle dead agents and might just pass their last observation which could also impact recurrent networks if entire sequences for an agent is the same observation and it trains on zero rewards etc. Additionally, if in a unique circumstance that a sequence is sampled where the first value in a sequence for a specific agent is that agents last timestep then it should learn on that and we wouldn't be able to use discounts as it would all be zero and we wouldn't know. This becomes very tricky to deal with without having a specific termination mask that would need to be in place in reverb most likely. When I used discounts to mask loss (even though incorrect and not training on the last action/value) it improved stability seemingly (not with ppo though). 

### Expected behavior

Some way to correctly get agent data validity masks.

### Possible Solution
Not obligatory, but suggest a fix/reason for the bug.

Require some environment information such as whether agents are dead or not and include it in reverb.

**Comment by OmaymaMahjoub** [comment]
Hello @EdanToledo, yeah we have this identical issue in #838 as we tried to make the loss masked and we ended up with this error:
`TrajectoryColumns cannot contain any None data references, got [<reverb.libpybind.WeakCellRef object at 0x7fe9a4308db0>, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None] for TrajectoryColumn at path ('next_extras', 'network_int_keys', 'agent_0', slice(-21, -1, None))`
Which what you have mentioned. However, we are currently working on it and soon I will reach you back when we confirm that we found a solution for it.

**Comment by DriesSmit** [comment]
Hey @EdanToledo :slightly_smiling_face: As @OmaymaMahjoub mentioned we started implementing loss masking in #838. However, there might still be a bug or two in the code. As you mentioned there are a lot of edge cases to the loss masking. So it is an ongoing issue to try and improve the PR. 
 Another issue is when we mask the critic in a SMAC environment. In this case, we want to calculate the correct returns as agents can still receive rewards after they die. Maybe check that PR and see if we can improve it to better accommodate your environment?
